src/testModel.ts

```diff
@@ -140,82 +107,65 @@ export class TestModel {
     return project;
   }
 
   private _updateProject(project: TestProject, projectReport: ProjectConfigWithFiles) {
     const filesToKeep = new Set<string>();
+    const files = projectFiles(project);
     for (const file of projectReport.files) {
       filesToKeep.add(file);
-      const testFile = project.files.get(file);
-      if (!testFile)
-        this._createFile(project, file);
+      const testFile = files.get(file);
+      if (!testFile) {
+        const testFile = new TeleSuite(file, 'file');
+        testFile.location = { file, line: 0, column: 0 };
+        files.set(file, testFile);
+      }
     }
 
-    for (const file of project.files.keys()) {
+    for (const file of files.keys()) {
       if (!filesToKeep.has(file))
-        project.files.delete(file);
+        files.delete(file);
     }
-  }
-
-  private _createFile(project: TestProject, file: string): TestFile {
-    const testFile = new TestFile(project, file);
-    project.files.set(file, testFile);
-    return testFile;
+    project.suite.suites = [...files.values()];
   }
 
   async workspaceChanged(change: WorkspaceChange) {
-    let modelChanged = false;
-    // Translate source maps from files to sources.
-    change.changed = this._mapFilesToSources(change.changed);
-    change.created = this._mapFilesToSources(change.created);
-    change.deleted = this._mapFilesToSources(change.deleted);
+    const testDirs = [...new Set([...this._projects.values()].map(p => p.project.testDir))];
 
-    if (change.deleted.size) {
-      for (const project of this._projects.values()) {
-        for (const file of change.deleted) {
-          if (project.files.has(file)) {
-            project.files.delete(file);
-            modelChanged = true;
-          }
-        }
-      }
-    }
-
-    if (change.created.size) {
-      let hasMatchingFiles = false;
-      for (const project of this._projects.values()) {
-        for (const file of change.created) {
-          if (file.startsWith(project.testDir))
-            hasMatchingFiles = true;
-        }
-      }
-      if (hasMatchingFiles)
-        await this.listFiles();
-    }
+    const changed = this._mapFilesToSources(testDirs, change.changed);
+    const created = this._mapFilesToSources(testDirs, change.created);
+    const deleted = this._mapFilesToSources(testDirs, change.deleted);
 
-    if (change.changed.size) {
-      const filesToLoad = new Set<string>();
-      for (const project of this._projects.values()) {
-        for (const file of change.changed) {
-          const testFile = project.files.get(file);
-          if (!testFile || !testFile.entries())
-            continue;
-          filesToLoad.add(file);
-        }
-      }
-      if (filesToLoad.size)
-        await this.listTests([...filesToLoad]);
-    }
-    if (modelChanged)
-      this._didUpdate.fire();
+    if (created.length || deleted.length)
+      await this.listFiles();
+    if (changed.length)
+      await this.listTests(changed);
```

Comment:
Documenting semantic change: when a file that has not listed the test yet is updated in the workspace, we force listing tests in it. Previously, we were not doing that. I hope that's intentional.

Reply:
Yes, intentional!